mythril/laser/ethereum/state.py

```diff
@@ -1,16 +1,44 @@
-from z3 import BitVec, BitVecVal, Solver, ExprRef, sat
+from z3 import BitVec, BitVecVal, Solver, ExprRef, Concat, sat, simplify
 from mythril.disassembler.disassembly import Disassembly
 from copy import copy, deepcopy
 from enum import Enum
 from random import randint
 
 from mythril.laser.ethereum.evm_exceptions import StackOverflowException, StackUnderflowException
 
-
 class CalldataType(Enum):
     CONCRETE = 1
     SYMBOLIC = 2
 
+class SymbolicCalldata:
+    def __init__(self, tx_id: int, starting_calldata: bytes=None):
+        self.tx_id = tx_id
+        self._calldata = {}
+
+        if starting_calldata:
+            for i in range(len(starting_calldata)):
+                self._calldata[i] = BitVecVal(starting_calldata[i], 8)
+
+    def get_word_at(self, index: int):
+        return self[index:index+32]
+
+    def __getitem__(self, item: int):
+        if isinstance(item, slice):
+            if item.step != None \
+            or item.start > item.stop \
+            or item.start < 0 \
+            or item.stop < 0: raise IndexError("Invalid Calldata Slice")
+
+            dataparts = []
+            for i in range(item.start, item.stop):
+                dataparts.append(self[i])
+            return simplify(Concat(dataparts))
+        else:
```

Comment:
This else isn't necessary, so to reduce nesting you can do the following:

```python3
def __getitem__(self, item: int):
    if isinstance(item, slice):
        if item.step != None \
            or item.start > item.stop \
            or item.start < 0 \
            or item.stop < 0: raise IndexError("Invalid Calldata Slice")

        dataparts = []
        for i in range(item.start, item.stop):
            dataparts.append(self[i])
        return simplify(Concat(dataparts))


    try:
        return self._calldata[item]
    except KeyError:
        self._calldata[item] = BitVec(str(self.tx_id) + '_calldata[' + str(item) + ']', 8)
        return self._calldata[item]
```

Reply:
hey @nbanmp, did you miss this comment, as i didn't see you responding to this :sweat_smile: .

Reply:
:sweat_smile: I misread something and thought that is was a non-issue anymore... Fixing.